### PR TITLE
feat: promisify dialog.showMessageBox() 

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Promisify `dialog.showMessageBox()` by splitting it into a sync and async method.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Split `dialog.showMessageBox()` into a synchronous version and a version that returns a Promise.

### CODE REVIEW

File: docs/api/dialog.md [hunk 1]
```diff
@@ -247,21 +247,61 @@ expanding and collapsing the dialog.
...
 It returns the index of the clicked button.
 
 The `browserWindow` argument allows the dialog to attach itself to a parent window, making it modal.
 
-If the `callback` and `browserWindow` arguments are passed, the dialog will not block the process. The API call
-will be asynchronous and the result will be passed via `callback(response)`.
+### `dialog.showMessageBox([browserWindow, ]options)`
+
+* `browserWindow` [BrowserWindow](browser-window.md) (optional)
+* `options` Object
```
Comment (MarshallOfSound) [inline docs/api/dialog.md]: As a future task we should find a way to de-dupe this documentation, having the same object documented twice is going to cause divergence issues in the future

File: atom/browser/api/atom_api_dialog.cc [hunk 2]
```diff
@@ -23,33 +23,59 @@
...
+                       const std::string& checkbox_label,
+                       bool checkbox_checked,
+                       const gfx::ImageSkia& icon,
+                       atom::NativeWindow* window) {
+  return atom::ShowMessageBoxSync(
+      window, static_cast<atom::MessageBoxType>(type), buttons, default_id,
+      cancel_id, options, title, message, detail, icon);
+}
+
+void ResolvePromiseObject(const atom::util::CopyablePromise& promise,
```
Comment (zcbenz) [inline atom/browser/api/atom_api_dialog.cc]: `CopyablePromise` should only be used when there is no other choice, such as passing it to some Chromium APIs.

For `ShowMessageBox`, we should change its callback type to `base::OnceCallback`, and then use `base::BindOnce` and move-only `Promise` for callbacks.

File: atom/browser/ui/message_box_win.cc [hunk 3]
```diff
@@ -220,28 +219,29 @@ void RunMessageBoxInNewThread(base::Thread* thread,
...
-                              const MessageBoxCallback& callback) {
+                              MessageBoxCallback callback) {
   int result = ShowTaskDialogUTF8(parent, type, buttons, default_id, cancel_id,
                                   options, title, message, detail,
                                   checkbox_label, &checkbox_checked, icon);
-  base::PostTaskWithTraits(FROM_HERE, {content::BrowserThread::UI},
-                           base::Bind(callback, result, checkbox_checked));
+  base::PostTaskWithTraits(
+      FROM_HERE, {content::BrowserThread::UI},
+      base::Bind(std::move(callback), result, checkbox_checked));
```
Comment (zcbenz) [inline atom/browser/ui/message_box_win.cc]: Should be `BindOnce`.

**Comment by audionerd** [comment]
This should probably be listed as a "breaking change" in the release notes.

Any existing call to `dialog.showMessageBox` will now return a Promise instead of a value. So users will need to change to `dialog.showMessageBoxSync` to return to the prior behavior, or modify their code to accept a Promise instead.

**Comment by mhenr18** [comment]
I fully support @audionerd 's comment about updating the docs - I only caught this because the typings for electron updated to reflect the breaking API change and our typescript codebase failed to compile.

**Pull request merged**